Add FastSpeech2Conformer

connor-henderson opened the pull request:
# What does this PR do?

Adds the TTS (text-to-speech) conformer version of the FastSpeech2 model. Closest related issue is [#15166](https://github.com/huggingface/transformers/issues/15166) though this implements ESPnet's conformer version instead of Fairseq's version as suggested in https://github.com/huggingface/transformers/pull/15773#issuecomment-1529558164.

[FastSpeech2 paper (Microsoft)](https://arxiv.org/pdf/2006.04558.pdf)
[Conformer version paper (ESPnet)](https://arxiv.org/pdf/2010.13956.pdf)

Conformer version code implementation: https://github.com/espnet/espnet/tree/master/espnet2/tts/fastspeech2
Additional conformer version code code implementation: https://github.com/DigitalPhonetics/IMS-Toucan/blob/ToucanTTS/TrainingInterfaces/Text_to_Spectrogram/FastSpeech2

The paper abstracts say most of this, but the main points of what makes this model an interesting addition are:
- It's non auto-regressive, leading to faster inference since it doesn't have to make predictions sequentially (hence the name `FastSpeech`)
- Uses a variance predictor in between the encoder and decoder to explicitly predict duration, pitch, and energy, leading to more accurate results
- Conformer architectures have been shown to improve performance in text to speech tasks, with the convolutions learning close range speech patterns and transformer attention helping with understanding longer range contexts
- There is currently only one other text-to-speech model in `transformers` (`SpeechT5ForTextToSpeech`)


## To do
- [x] Prepared 🤗 Transformers dev environment
- [x] Set up debugging environment of the original repository
- [x] Created script that successfully runs the forward() pass using the original repository and checkpoint
- [x] Successfully added the model skeleton to 🤗 Transformers (+ vocoder)
- [x] Successfully converted original checkpoint to 🤗 Transformers checkpoint (+ vocoder)
- [x] Successfully ran forward() pass in 🤗 Transformers that gives identical output to original checkpoint (+ vocoder)
- [x] Finished model tests in 🤗 Transformers
- [x] Successfully added tokenizer in 🤗 Transformers
- [x] Run end-to-end integration tests
- [x] Finished docs
- [x] Uploaded model weights to the Hub (will ask they're moved to just `fastspeech2_conformer` when ready)

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer: @stas00, Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->
@hollance @sanchit-gandhi

Commit:
start - docs, SpeechT5 copy and rename

Commit:
add relevant code from FastSpeech2 draft, have tests pass

Commit:
make it an actual conformer, demo ex.

Commit:
matching inference with original repo, includes debug code

Commit:
refactor nn.Sequentials, start more desc. var names

Commit:
more renaming

Commit:
more renaming

Commit:
vocoder scratchwork

Commit:
matching vocoder outputs

Commit:
hifigan vocoder conversion script

Commit:
convert model script, rename some config vars

Commit:
replace postnet with speecht5's implementation

Commit:
passing common tests, file cleanup

Commit:
expand testing, add output hidden states and attention

Commit:
tokenizer + passing tokenizer tests

Commit:
variety of updates and tests

Commit:
g2p_en pckg setup

Commit:
import structure edits

Commit:
docstrings and cleanup

Commit:
repo consistency

Commit:
deps

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_23439). All of your documentation changes will be reflected on that endpoint.

Commit:
small cleanup

Commit:
forward signature param order

hollance reviewed (changes requested):
Hey Connor, apologies for the delay in doing the review. Aside from the handful of small things below, I think this is an excellent PR already. 

I can understand some of the design decisions you made, but the Transformers team is pretty strict about how the code should be structured, and so some of this code will have to change in order to be accepted. I don't think it's going to be too much work, though, since what you have here is pretty solid. 😄

Commit:
address comments except for masks and labels

Commit:
address comments on attention_mask and labels

connor-henderson commented on setup.py:
```diff
@@ -118,6 +118,7 @@
     "flax>=0.4.1,<=0.6.9",
     "ftfy",
     "fugashi>=1.0",
+    "g2p-en>=2.1.0",
```
"Just" the g2p_en.G2p().__call__ method, and it's only called within `_tokenize`. Its fairly nested and I pulled it out into a notebook to illustrate what we'd be re-implementing [here](https://colab.research.google.com/drive/1u4utpaYgthPZ3rXmdeucTYB-8r81SUPe). It's a couple hundred lines and  notably still requires packages like nltk and doesn't include the long-ish data from the `homograph.en` file used to create the homograph2features dictionary. The amount of work / code to reimplement it would also depend on how stringently we choose to match the outputs to the original.

Imo it's best to simply keep `g2p_en`, but it's hard for me to evaluate the tradeoff without knowing more about how the maintainers view deps like this.

connor-henderson commented on src/transformers/models/fastspeech2_conformer/configuration_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,290 @@
+
+
+FASTSPEECH2_CONFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "connor-henderson/fastspeech2_conformer": "https://huggingface.co/connor-henderson/fastspeech2_conformer/raw/main/config.json",
```
I think when they looked at this for the Fairseq version of FastSpeech2 they opted for no organization ([comment](https://github.com/huggingface/transformers/pull/15773#discussion_r852197361))? Good with whatever

connor-henderson commented on src/transformers/models/fastspeech2_conformer/configuration_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,290 @@
+        encoder_concat_after=False,
+        decoder_concat_after=False,
+        reduction_factor=1,
+        # encoder / decoder
```
Sg, removed

connor-henderson commented on src/transformers/models/fastspeech2_conformer/configuration_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,290 @@
+    @property
+    def var_pred_num_bins(self):
+        # Number of bins in the variance predictors.
+        return 256
```
Realized these were actually never used and are captured by other config vars anyway, so I removed them

connor-henderson commented on src/transformers/models/fastspeech2_conformer/tokenization_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,175 @@
+        **kwargs,
+    ):
+        self.init_backend()
+        import g2p_en
```
Updated and opted to copy the pattern in `tokenization_wav2vec2_phoneme.py` for when it imports `phonemizer`, lmk if a different one is preferred

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1939 @@
+        return hidden_states + layer_output.transpose(1, 2)
+
+
+class FastSpeech2ConformerLayerNorm(torch.nn.LayerNorm):
```
Sounds good, made it a module

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1939 @@
+        kernel_size=3,
+        bias=True,
+        dropout_rate=0.5,
+    ):
```
The variance predictor is instantiated twice with two different sets of args passed in, I don't see a way to pass in the config without significantly re-organizing the code. Lmk if that's still preferred

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
````diff
@@ -0,0 +1,1939 @@
+
+        Example:
+
+        ```python
+        >>> from transformers import (
+        ...     FastSpeech2ConformerTokenizer,
+        ...     FastSpeech2ConformerModel,
+        ...     FastSpeech2ConformerHifiGan,
+        ... )
+
+        >>> tokenizer = FastSpeech2ConformerTokenizer.from_pretrained("connor-henderson/fastspeech2_conformer")
+        >>> inputs = tokenizer("some text to convert to speech", return_tensors="pt")
+        >>> input_ids = inputs["input_ids"]
+
+        >>> model = FastSpeech2ConformerModel.from_pretrained("connor-henderson/fastspeech2_conformer")
+        >>> output_dict = model(input_ids, return_dict=True)
+        >>> spectrogram = output_dict["spectrogram"]
+
+        >>> vocoder = FastSpeech2ConformerHifiGan.from_pretrained("connor-henderson/fastspeech2_conformer_hifigan")
+        >>> waveform = vocoder(spectrogram)
+        >>> print(waveform.shape)
+        torch.Size([1, 49664])
+        ```
````
Oh cool, didn't know about that! Ran it and it passed

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1939 @@
+        n_head (int): The number of heads.
+        n_feat (int): The number of features.
+        dropout_rate (float): Dropout rate.
+    """
```
Ended up getting rid of this class. The `forward_qkv` and `forward_attention` methods are the only ones needed for the `RelPositionMultiHeadedAttention` so I just pulling those methods + the variable inits in to that class.

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1939 @@
+        input_ids: torch.LongTensor,
+        input_lengths: Optional[torch.LongTensor] = None,
+        target_spectrograms: Optional[torch.FloatTensor] = None,
+        spectrogram_lengths: Optional[torch.LongTensor] = None,
```
Updated the model and tokenizer to use `attention_mask` instead.

Re `labels`, FastSpeech2 is somewhat unique in that it takes in many labels (spectrograms, pitch, energy, and duration) for training. I'm not sure exactly what this means for compatibility with `Trainer` since I haven't had time to do a deeper dive, but for now I've changed the "targets" to include `_labels` in their name, left the training test as skipped, and planning to look into it more when I do the demo notebook.

hollance replied:
AFAIK, Trainer will use any input that has `labels` in its name as the targets, so to have multiple targets you'd indeed have multiple arguments named `something_labels`.

connor-henderson commented:
Thanks for the review @hollance! Addressed the comments above, the only part that might need follow-up discussion is making the `labels` compatible with the `Trainer`
>Re labels, FastSpeech2 is somewhat unique in that it takes in many labels (spectrograms, pitch, energy, and duration) for training. I'm not sure exactly what this means for compatibility with Trainer since I haven't had time to do a deeper dive, but for now I've changed the "targets" to include _labels in their name, left the training test as skipped, and planning to look into it more when I do the demo notebook.

hollance commented on src/transformers/models/fastspeech2_conformer/configuration_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,289 @@
+    documentation from [`PretrainedConfig`] for more information.
+
+
+    Args:
```
Is the documentation for the arguments missing here?

connor-henderson replied:
Yes thanks, added

hollance commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1754 @@
+        return self.dropout(hidden_states), self.dropout(pos_emb)
+
+
+class FastSpeech2ConformerEncoder(nn.Module):
```
The order of the classes in Transformers modeling files is usually: If class B uses class A, then class A goes first in the file. You've kind of ordered them the other way around. Not a big deal, but we generally order things so that if something is used, you've already seen its definition.

connor-henderson replied:
Reordered to match this pattern

hollance commented on src/transformers/models/speecht5/configuration_speecht5.py:
```diff
@@ -359,8 +359,6 @@ class SpeechT5HifiGanConfig(PretrainedConfig):
     Args:
         model_in_dim (`int`, *optional*, defaults to 80):
             The number of frequency bins in the input log-mel spectrogram.
-        sampling_rate (`int`, *optional*, defaults to 16000):
-            The sampling rate at which the output audio will be generated, expressed in hertz (Hz).
```
Not sure why this was deleted?

connor-henderson replied:
Noticed it wasn't used, but just put it back in the interest of keeping this PR to just touching one model

hollance commented on tests/models/fastspeech2_conformer/test_modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,361 @@
+# # # # # # # # # # # # # # # # # # # # # # #
+#  WIP - Currently has old draft FS2 tests  #
+# # # # # # # # # # # # # # # # # # # # # # #
```
Is this still the case or an old comment?

connor-henderson replied:
Old comment, removed

hollance reviewed (commented):
Hey Connor, just a few more comments on the code. It's looking very good already!

If you're happy with the code so far, feel free to request a review from @sanchit-gandhi and/or @ArthurZucker.

Commit:
address second round of comments

Commit:
remove old unneeded line

connor-henderson commented:
Appreciate the comments @hollance, @ArthurZucker @sanchit-gandhi this should be ready for review for you now

sanchit-gandhi commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+            # remove seq-len dim since this collapses to 1
+            waveform = hidden_states.squeeze(1)
+
+        return waveform
```
Would be nice to have a model called `FastSpeech2ConformerWithHiFiGan` that uses the FS2 model as a base model and the HiFi Gan as a head on top -> this way a user only has to do `from_pretrained` once to get the weights of both models and the vocoder is used directly in the forward pass (better API IMO)

connor-henderson replied:
To my knowledge this would be the first model in the library to have a head that is itself a self-contained model in the library. 

Given this, I’m not certain what’s best to do for the config for this combined model. If we have an entirely separate config that combines attributes of both the model and vocoder, we’ll have to update two separate configs when changes are made.

The only reason the vocoder is separate in the first place is to follow the `SpeechT5` pattern and since different vocoders could theoretically be used, but maybe we should just move it into the base model and determine whether to pass the outputs through the vocoder using a config attribute or forward pass arg?

ylacombe replied:
Hi @connor-henderson, 
`SpeechT5` pattern is not the typical transformers pattern with the `generate_speech` method, and I think we're moving towards a more consistent library so I guess we should not strictly copy `SpeechT5` pattern.

There are at least two models that I know of, [MusicGen](https://github.com/huggingface/transformers/pull/24109) and [Bark](https://github.com/huggingface/transformers/pull/24086), that uses self-containted models as head.

You can get inspiration from both models for properly initiating the configurations and the models!

connor-henderson replied:
Ah thanks @ylacombe, hadn't seen those new models yet the `BarkConfig` with `from_sub_model_configs` is exactly what I was looking for

sanchit-gandhi commented on src/transformers/models/fastspeech2_conformer/tokenization_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,176 @@
+
+    def init_backend(self):
+        """Initializes the backend."""
+        requires_backends(self, "g2p_en")
```
Nice!

sanchit-gandhi commented on tests/models/fastspeech2_conformer/test_modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,356 @@
+@require_torch
+@require_g2p_en
+@slow
+class FastSpeech2ConformerModelIntegrationTest(unittest.TestCase):
```
Worth doing a slow test to make sure we get the right loss out too? Checks that we have the non-log domain parts correct, and the loss terms too

connor-henderson replied:
Added a slow test for the loss including a comment about labels for training / calculating loss. Didn't write a feature extractor since I thought the dependencies for duration extraction (montreal forced aligner) and pitch and energy extraction (librosa) might be unwanted

sanchit-gandhi replied:
Good with me to add with just inference first

sanchit-gandhi commented on tests/models/fastspeech2_conformer/test_tokenization_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,166 @@
+        self.assertListEqual(tokenizer.convert_ids_to_tokens(ids), tokens)
+
+    @slow
+    def test_tokenizer_integration(self):
```
Can we check the niche aspects of this tokenizer? E.g. do the properties in `_clean_text` work as expected?

connor-henderson replied:
The `_clean_text` method is called within tokenize and so is implicitly tested in the `test_tokenizer_integration` test and its also the only somewhat niche method on this tokenizer. Did you have something in particular in mind?

sanchit-gandhi reviewed (commented):
Looks great already @connor-henderson! Mainly just refactoring / styling suggestions from my part. Would be great to bring the attention / transformer layers into alignment a bit more with the rest of the library (see suggestions below), and think we can simplify a lot of the `__init__` logic by always passing the `config` as an attribute and avoiding nested configs.

Happy to clarify any of the points below or answer any questions! Also cc'ing @ylacombe

sanchit-gandhi reviewed (commented):
Looks great already @connor-henderson! Mainly just refactoring / styling suggestions from my part. Would be great to bring the attention / transformer layers into alignment a bit more with the rest of the library (see suggestions below), and think we can simplify a lot of the `__init__` logic by always passing the `config` as an attribute and avoiding nested configs.

Happy to clarify any of the points below or answer any questions! Also cc'ing @ylacombe

Commit:
address comments part 1

connor-henderson commented on src/transformers/models/fastspeech2_conformer/configuration_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,289 @@
+        hidden_size=384,
+        input_dim=78,
+        num_mel_bins=80,
+        num_attention_heads=2,
```
Edited to have separate arg for each

sanchit-gandhi replied:
Thank you!

connor-henderson commented on src/transformers/models/fastspeech2_conformer/configuration_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,289 @@
+        self.speech_decoder_postnet_dropout = speech_decoder_postnet_dropout
+        self.speech_decoder_postnet_kernel = speech_decoder_postnet_kernel
+        self.speech_decoder_postnet_layers = speech_decoder_postnet_layers
+        self.should_postnet_compute_logits = False
```
Had this as a workaround so the postnet could be `Copied from` SpeechT5's, but per your below comment about their making them distinct I've removed this

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+    """
+    Duration predictor module.
+
+    This is a module of duration predictor described in `FastSpeech: Fast, Robust and Controllable Text to Speech`_.
```
Sure, updated

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+        super().__init__()
+
+        self.conv_layers = torch.nn.ModuleList()
+        self.log_domain_offset = 1.0
```
Yea odd, its not configurable in the original espnet library

sanchit-gandhi replied:
Let's leave as is then

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+    def forward(self, encoder_hidden_states, padding_masks=None, is_inference=False):
+        """
+        Args:
+            hidden_states (Tensor): Batch of input sequences (batch_size, max_text_length, input_dim).
```
Updated them to match

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+class FastSpeech2ConformerDimensionalLayerNorm(nn.Module):
+    def __init__(self, output_dim, dim=-1):
+        super().__init__()
+        self.layer_norm = nn.LayerNorm(output_dim, eps=1e-12)
```
Should be fine to use the default, doesn’t impact the integration test and I’d just used 1e-12 to mimic the espnet lib’s usage of it. Deleted so we use the default eps of 1e-5.

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+
+
+class FastSpeech2ConformerDimensionalLayerNorm(nn.Module):
+    def __init__(self, output_dim, dim=-1):
```
I made the kwarg default `dim=1` and stopped passing one in, does that work for you? Happy to just hard code it if you want but I’m not sure what addt’l benefit that would give us

sanchit-gandhi replied:
Sounds good

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+
+        return pos_tensor
+
+    def forward(self, query, key, value, pos_emb, mask):
```
Did a refactor to bring it more into alignment with the ones FS2 has, but note that FS2 does not use ones like `past_key_value` since its non auto-regressive

sanchit-gandhi replied:
Perfect - that's great thanks

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+        self.pos_bias_u = nn.Parameter(torch.Tensor(self.num_attention_heads, self.dim_key))
+        self.pos_bias_v = nn.Parameter(torch.Tensor(self.num_attention_heads, self.dim_key))
+
+    def forward_qkv(self, query, key, value):
```
Moved to the forward method

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+            groups=channels,
+            bias=bias,
+        )
+        self.norm = nn.BatchNorm1d(channels)
```
Yes, removed

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+        return self.conv2(self.dropout(hidden_states).transpose(-1, 1)).transpose(-1, 1)
+
+
+class FastSpeech2ConformerRelPositionalEncoding(nn.Module):
```
`Wav2Vec2ConformerRelPositionalEmbedding` is slightly different and causes a ~1e-2 divergence, think it has to do with whether the positive embeddings are flipped

sanchit-gandhi replied:
Sounds good - let's leave as is

connor-henderson replied:
Cool this works, replaced it with Llama's

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+    ):
+        super().__init__()
+
+        self.self_attn = self_attn
```
Moved them so we don't have to pass nn layers through

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+            last_hidden_state=hidden_states, hidden_states=all_hidden_states, attentions=all_self_attentions
+        )
+
+    def _integrate_with_utt_embed(self, hidden_states, utt_embeddings):
```
Moved to the forward

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+        self.input_dim = config.input_dim
+        self.num_mel_bins = config.num_mel_bins
+        self.hidden_size = config.hidden_size
+        self.eos_token_id = self.input_dim - 1
```
On further review we actually don't need the eos, bos, or pad token ids in the config, removing them all

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+                Batch of padded token-averaged pitch.
+            energy_labels (`torch.FloatTensor` of shape `(batch_size, sequence_length + 1, 1)`, *optional*):
+                Batch of padded token-averaged energy.
+            utterance_embedding (`torch.FloatTensor` of shape `(batch_size, embedding_dim)`, *optional*):
```
Yea renamed

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+                Tensor containing the utterance embeddings.
+            alpha (`float`, *optional*):
+                Alpha to control the speed of the spectrogram generation. Default is `1.0`.
+            lang_id (`torch.LongTensor`, *optional*):
```
Renamed to lang_ids since the name is misleading, it actually has to be passed in the forward because it is different for different batches of audio. Below is the configuration var relevant to it
```py
num_languages (`int`, *optional*, defaults to `None`):
            Number of languages. If set to > 1, assume that the language ids will be provided as the input and use the
            languge id embedding layer.
```

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+                attention_mask=attention_mask,
+                out_mask=spectrogram_mask,
+            )
+            loss = l1_loss + duration_loss + pitch_loss + energy_loss
```
Sure, summed in there instead

Commit:
address comments pt 2

Commit:
rename auto mapping

Commit:
fixes for failing tests

Commit:
address comments part 3 (bart-like, train loss)

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1781 @@
+        if self.concat_after:
+            self.concat_linear = nn.Linear(size + size, size)
+
+    def forward(self, hidden_states, pos_emb, mask):
```
Updated them to match BART

Commit:
make style

Commit:
pass config where possible

connor-henderson commented:
Thank you for the review @sanchit-gandhi, comments should be addressed now. 

Centralizing a note on passing the config instead of args here since there were a few comments on that - the other modules mentioned are all instantiated twice with different arg values so they can’t solely be passed the config.  Lmk if you think there’s something I missed or if you’d prefer something else like duplicating the modules in order to pass just the config.

sanchit-gandhi commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1693 @@
+        self.config = config
+
+    @torch.no_grad()
+    def generate_speech(self, input_ids: torch.LongTensor, **kwargs) -> torch.FloatTensor:
```
IMO a bit misleading to have `generate` in the name here when really we're just doing a forward pass. Would instead define a forward method and have this take as argument all the same arguments as `FastSpeech2ConformerModel`:
```python
    def forward(self, input_ids: torch.LongTensor, attention_mask=...) -> Union[Tuple, FastSpeech2ConformerWithHiFiGanModelOutput]:
```

sanchit-gandhi commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1693 @@
+        Args:
+            input_ids (`torch.LongTensor` of shape `(batch_size, sequence_length)`):
+                Input sequence of text vectors.
+            kwargs:
```
Let's write them out explicitly please - easier for the user and only one copy-and-paste for us!

sanchit-gandhi commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1693 @@
+        output_dict = self.model(input_ids, **kwargs, return_dict=True)
+        spectrogram = output_dict["spectrogram"]
+        waveform = self.vocoder(spectrogram)
+        return waveform
```
Here can we return a `ModelOutput` object with all the same entries as `FastSpeech2ConformerModelOutput` **and** a new one for `waveform`?

sanchit-gandhi commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
````diff
@@ -0,0 +1,1693 @@
+        torch.Size([1, 49664])
+        ```
+        """
+        output_dict = self.model(input_ids, **kwargs, return_dict=True)
````
Let's not force this argument `return_dict=True` but rather let the user input it as an argument. We'll return either a tuple or `FastSpeech2ConformerModelOutput` from the forward call

sanchit-gandhi commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1693 @@
+                Arguments to be passed to the forward call of `FastSpeech2ConformerModel` model-specific kwargs that
+                will be forwarded to the `forward` function of the model.
+
+        Example:
```
I like this API -> think this is the best way of integrating a vocoder into a TTS model (cc @ylacombe)

ylacombe replied:
Totally agree, especially if users can change the vocoder if they want!

connor-henderson replied:
@sanchit-gandhi did you mean update `SpeechT5`, or is there something else wanted here?

sanchit-gandhi replied:
No no all good here! Just wanted to highlight that this API is better than the SpeechT5 one, and that I think we can update the SpeechT5 one to match this! Updated my original comment to clarify!

sanchit-gandhi commented on src/transformers/models/fastspeech2_conformer/tokenization_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,176 @@
+}
+
+
+class FastSpeech2ConformerTokenizer(PreTrainedTokenizer):
```
LGTM - @ArthurZucker would you mind casting your eye over the tokenizer when you get the chance? Thanks!

ArthurZucker replied:
Sure!

sanchit-gandhi replied:
And more generally the PR when you get the chance, this is ready for a core maintainers' review :)

sanchit-gandhi commented on src/transformers/models/fastspeech2_conformer/tokenization_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,176 @@
+    """
+    Construct a FastSpeech2Conformer tokenizer.
+
+    Args:
```
A few missing args from the docstring

connor-henderson replied:
Thanks added

sanchit-gandhi commented on tests/models/fastspeech2_conformer/test_modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,468 @@
+        self.assertTrue(torch.allclose(spectrogram[0, :10, :10], expected_mel_spectrogram, atol=1e-4))
+        self.assertEqual(spectrogram.shape, (1, 205, model.config.num_mel_bins))
+
+    def test_training_integration(self):
```
Nice test

sanchit-gandhi commented on tests/models/fastspeech2_conformer/test_modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,468 @@
+
+@require_torch
+class FastSpeech2ConformerModelTest(ModelTesterMixin, unittest.TestCase):
+    all_model_classes = (FastSpeech2ConformerModel,) if is_torch_available() else ()
```
We can test the `FastSpeech2ConformerWithHiFiGanModel` if we define it with a forward pass

connor-henderson replied:
Added a forward pass and separate test class for that model since it uses a different config and has different expected output len

sanchit-gandhi reviewed (commented):
Thanks for iterating here @connor-henderson, the PR is looking really good! Just a few small suggestions, then I think we can get a final review here and move the checkpoints to the official org

Commit:
add forward method + tests to WithHifiGan model

Commit:
make style

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1693 @@
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, encoded_embeddings, duration_labels, speaking_speed=1.0):
```
Yea it's actually not an arg in the model's forward method, the value comes from the config, but I believe we still want to pass it to the length regulator's forward so it can still be configured easily after the model is instantiated. Updated it to be updated by the `config.speaking_speed` instead of the `model.speaking_speed`.

Are you also saying you'd prefer to send the whole config on the forward call?

(from the model forward)
```py
hidden_states = self.length_regulator(hidden_states, duration_predictions, self.config.speaking_speed)
```

sanchit-gandhi replied:
I was thinking that we just set it in the `__init__` of the length regulator directly from the config:
```python
    def __init__(self, config):
        super().__init__()
        self.speaking_speed = config.speaking_speed
```

I'm not sure how much the `speaking_speed` attribute would actually be changed in practice, which is why I think it's fine just to set once in the config and not worry about updating it.

If we disagree with this, and want to let the user configure the speaking speed, then let's keep the `speaking_speed` argument to the length regulator, and add a line in `FastSpeech2ConformerModel`:
```python
speaking_speed = speaking_speed if speaking_speed is not None else self.config.speaking_speed
```

cc @Vaibhavs10 @ylacombe

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1693 @@
+
+
+class FastSpeech2ConformerConvolutionModule(nn.Module):
+    def __init__(self, channels, kernel_size, bias=True):
```
For sure I'll remove the bias arg (it's always True) and the channels arg (and pass the config instead to access the config.hidden_size), but the kernel size param varies between the encoder and decoder (`encoder_kernel_size` and `decoder_kernel_size`) so I still think we need to pass that through, no?

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1693 @@
+
+
+class FastSpeech2ConformerVariancePredictorLayer(nn.Module):
+    def __init__(self, input_channels, num_chans, kernel_size, dropout_rate, bias):
```
The bias is fixed as True so I have now removed that arg. As for the others, they come from the below config values and I'm not sure how we'd disambiguate which to use without passing them through?

`energy_predictor_channels`
`energy_predictor_kernel_size`
`energy_predictor_dropout`

`pitch_predictor_channels`
`pitch_predictor_kernel_size`
`pitch_predictor_dropout`

sanchit-gandhi replied:
Okay good with me to keep them configurable then!

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1693 @@
+
+        self.duration_predictor = FastSpeech2ConformerDurationPredictor(config)
+
+        self.pitch_predictor = FastSpeech2ConformerVariancePredictor(
```
Same comment as above + the `pitch_predictor_layers` and `energy_predictor_layers` args make it hard to just pass in the config. I could pass in a string "energy" or "pitch" and use that to extract the right values from the config if that's preferred?

ArthurZucker replied:
I think it's fine for 4 arguments this way!

Commit:
address arg passing and generate_speech comments

sanchit-gandhi commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1810 @@
+    "The FastSpeech2ConformerModel with a FastSpeech2ConformerHifiGan vocoder head that performs text-to-speech (waveform).",
+    FASTSPEECH2_CONFORMER_WITH_HIFIGAN_START_DOCSTRING,
+)
+class FastSpeech2ConformerWithHifiGan(PreTrainedModel):
```
Very nice 👌

sanchit-gandhi reviewed (approved):
This is looking super clean - thanks for iterating @connor-henderson! Especially like the new "WithHiFiGanHead" class and overall the FS2 API. I think we've got as many modules as possible to be intialised from the `config` -> the remainder need more flexibility (as mentioned in the comments above). 

Would love a second opinion from @ArthurZucker on the tokenizer, and more generally on the PR integration before we get this merged!

ArthurZucker commented:
Reviewing now!

ArthurZucker commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1810 @@
+        num_chans = config.duration_predictor_channels
+        kernel_size = config.duration_predictor_kernel_size
+        input_channels = config.hidden_size if layer_idx == 0 else num_chans
+        self.layer = nn.ModuleList(
+            [
+                nn.Conv1d(
+                    input_channels,
+                    num_chans,
+                    kernel_size,
+                    stride=1,
+                    padding=(kernel_size - 1) // 2,
+                ),
+                nn.ReLU(),
+                FastSpeech2ConformerDimensionalLayerNorm(num_chans),
+                nn.Dropout(config.duration_predictor_dropout_rate),
+            ]
+        )
```
Why use ModuleList when you can just define `self.conv, self.act_fn, self.norm, self.dropout`?

ArthurZucker commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1810 @@
+                Batch of input sequences.
+            padding_masks (`torch.ByteTensor` of shape `(batch_size, max_text_length)`, *optional*):
+                Batch of masks indicating padded part.
+            is_inference (`bool`, *optional*): Whether or not the model is running inference.
```
```suggestion
            is_inference (`bool`, *optional*, defaults to `False`): Whether or not the model is running inference.
```

ArthurZucker commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1810 @@
+class FastSpeech2ConformerVariancePredictorLayer(nn.Module):
+    def __init__(self, input_channels, num_chans, kernel_size, dropout_rate):
+        super().__init__()
+        self.layer = nn.ModuleList(
+            [
+                nn.Conv1d(
+                    input_channels,
+                    num_chans,
+                    kernel_size,
+                    stride=1,
+                    padding=(kernel_size - 1) // 2,
+                    bias=True,
+                ),
+                nn.ReLU(),
+                FastSpeech2ConformerDimensionalLayerNorm(num_chans),
+                nn.Dropout(dropout_rate),
+            ]
+        )
+
```
is there a reason you are not simply using `self.conv1, self.act_fn, self.layer_norm` and remove the need for `FastSpeech2ConformerDimensionalLayerNorm`, and improve clarity?

ArthurZucker commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1810 @@
+            `torch.Tensor`: Output tensor of shape `(batch, time, size)`.
+
+        """
+        # whether to use macaron style
```
Seems like there is a lot of `if xxx`, are they all used by the released models? (Meaning if it's only a single pass that is always use, let's remove the complicated logic)

connor-henderson replied:
Microsoft didn't release their weights, but these if's are all configurable in the library with the implementation this used (espnet). We could remove the logic but it might break converting some that have been trained in espnet

ArthurZucker commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1810 @@
+        Returns:
+            torch.Tensor: Batch of output tensors (batch_size, T, hidden_channels).
+        """
+        hidden_states = torch.relu(self.conv1(hidden_states.transpose(-1, 1))).transpose(-1, 1)
+        return self.conv2(self.dropout(hidden_states).transpose(-1, 1)).transpose(-1, 1)
```
can you split this in more lines for readability?

ArthurZucker commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1810 @@
+    FastSpeech2ConformerEncoder encoder module.
+
+    Args:
+        config (`FastSpeech2ConformerConfig`): FastSpeech2ConformerConfig instance.
+        linear_units (`int`, *optional*, defaults to 2048): The number of units of position-wise feed forward.
+        num_blocks (`int`, *optional*, defaults to 6): The number of decoder blocks.
+        dropout_rate (`float`, *optional*, defaults to 0.1): Dropout rate.
+        positional_dropout_rate (`float`, *optional*, defaults to 0.1): Dropout rate after adding positional encoding.
+        attention_dropout_rate (`float`, *optional*, defaults to 0.0): Dropout rate in attention.
+        use_encoderinput_layer (`bool`, *optional*, defaults to `False`): Input layer type.
+        normalize_before (`bool`, *optional*, defaults to `True`): Whether to use layer_norm before the first block.
+        concat_after (`bool`, *optional*, defaults to `False`): Whether to concat attention layer's input and output.
+            if True, additional linear will be applied. i.e. x -> x + linear(concat(x, att(x))) if False, no additional
+            linear will be applied. i.e. x -> x + att(x)
+        cnn_module_kernel (int, *optional*, defaults to 31): Kernel size of convolution module.
```
Format just misses extra lines after the argument

ArthurZucker commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1810 @@
+    Transformers.
+    """
+
+    def __init__(self, config: FastSpeech2ConformerConfig):
```
Regarding all the messy config kwargs that have to be passed down, don't you think we should just have an `encoder_config` and `decoder_config` in the `FastSpeech2ConformerConfig` especially if there are many arguments that are different, but both use the same architecture. Would simplify things a lot!

connor-henderson replied:
@ArthurZucker is nested configs an acceptable pattern? I've only seen it for a couple of the newer audio models that combine >1 models into one like `FastSpeech2ConformerWithHifiGan`. This would also lead to `FastSpeech2ConformerWithHifiGan` being nested two layers deep. 

Happy to make the change if it's wanted

ArthurZucker replied:
In this case I think it makes sense. You can try to only have one layer of nesting if possible, but otherwise should be alright! I Think that it will end up being a more readable `config.json` 😉

sanchit-gandhi replied:
We can still probably maintain one layer of nesting by having the `FastSpeech2ConformerWithHifiGanConfig` have three arguments:
1. Encoder config
2. Decoder config
3. Vocoder config

=> this way we keep the 1-layer deep nesting

connor-henderson replied:
We would need model config, encoder config, decoder config, and vocoder config right? There's a number of attributes the model needs that aren't specific to encoder or decoder

sanchit-gandhi replied:
Looking very clean after the refactor

ArthurZucker reviewed (commented):
It already looks nice, great work for a hard model to add! Will do another pass once you adress current comments! Mostly nits on the modeling for readability, more comments on the tokenizer to remove the hard dependency! 
Very nice tests 🚀 🤗

Commit:
address Arthur comments

connor-henderson commented:
Thanks @ArthurZucker for the review, your comments should be addressed now

connor-henderson commented on src/transformers/models/auto/tokenization_auto.py:
@ArthurZucker per the failing doctest, are we just not able to have docstring examples in the model using the tokenizer if it a required package isn't a hard dependency? Looks like BioGpt doesn't have them

ArthurZucker replied:
Mmmmm we should be. This kind of goes in the `custom/exotic` tokenizer, which have different tests, but basically adding the package to the CI job should do the trick. 
I am not sure we want tokenizers like this to be tested with all the rest cc @ydshieh the king of tests 😉

ydshieh replied:
If here we are talking about doctest, adding the package to `doc_test_job = CircleCIJob( ...` in `.circleci/create_circleci_config.py` works for me.

sanchit-gandhi commented:
Looking very clean, thanks for iterating @connor-henderson! Would be great to get a look from @ylacombe to check this model will be compatible with the new TTS pipeline (e.g. with the auto mapping names), alongside @ArthurZucker for the final review

ylacombe commented on src/transformers/modeling_outputs.py:
```diff
@@ -1412,6 +1412,51 @@ class Seq2SeqSpectrogramOutput(ModelOutput):
+@dataclass
+class Seq2SeqSpectrogramOutputWithPastAndCrossAttentions(ModelOutput):
+    """
+    Base class for sequence-to-sequence spectrogram outputs.
```
(nit) the docstrings is not up-to-date

ylacombe commented on utils/check_repo.py:
```diff
@@ -83,6 +83,9 @@
     "ErnieMForInformationExtraction",
+    "FastSpeech2ConformerHifiGan",  # Already tested by SpeechT5HifiGan (# Copied from)
+    "FastSpeech2ConformerWithHifiGan",  # Built with two smaller (tested) models.
+    "GraphormerEncoder",  # Building part of bigger (tested) model.
```
This is probably nothing but why does `GraphormerEncoder` appears added in your PR ?

connor-henderson replied:
Yep it's from a rebase, removed

ylacombe commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1759 @@
+        if not return_dict:
+            return model_outputs + (waveform,)
+
+        return FastSpeech2ConformerWithHifiGanOutput(**model_outputs, waveform=waveform)
```
This should probably returns a waveform first IMO, or even the waveform only, I'll discuss that in details in the review!

sanchit-gandhi replied:
This is more or less the same as what we have with the Vits output, see:
https://github.com/huggingface/transformers/blob/866df66fe4a3b3b948b926d849ef291675b8a24c/src/transformers/models/vits/modeling_vits.py#L52
and:
https://github.com/huggingface/transformers/blob/866df66fe4a3b3b948b926d849ef291675b8a24c/src/transformers/models/vits/modeling_vits.py#L1505

=> I agree that we should try and be consistent here with the ordering

ylacombe replied:
Ah yes, it makes total sense @sanchit-gandhi, thanks for pointing the difference between autoencoder and autoregressive outputs.

Let's make sure that the waveform and the spectrogram comes first then.

ylacombe reviewed (approved):
Nice work @connor-henderson ! 

I'm just leaving some small comments and raise the following discussion:

cc @sanchit-gandhi, @ArthurZucker and @Narsil who might be interested

It's an opportunity to discuss what we want text-to-spectrograms and text-to-waveforms to return. Currently, and if I'm not mistaken, all TTS/TTA models supported by the library return the spectrogram or waveform directly, by default. 

Here @connor-henderson proposes to return a different output, more in line with the library traditional output. I'm starting to think that this should be what we're aiming to output by default for TTS models, i.e dictionaries, that could also support additional information, such as the waveform/spectrogram lengths if there is batching.

So there are two possibilities here:

1. Refactor a bit your code @connor-henderson , so that it returns by default the spectrogram/waveform only and update that in another PR that would take care of every TTS models
2. Leave your code as it is for now including **not** adding the models to the text-to-XXX auto mapping - and we'll update the auto-mapping once we agree on the default output.

In any case, it's a decision we need to make quickly and stick to for a while, to remain consistent!

ArthurZucker commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1759 @@
+        super().__init__()
+        self.leaky_relu_slope = leaky_relu_slope
+
+        self.convs1 = nn.ModuleList(
+            [
+                nn.Conv1d(
+                    channels,
+                    channels,
+                    kernel_size,
+                    stride=1,
+                    dilation=dilation[i],
+                    padding=self.get_padding(kernel_size, dilation[i]),
+                )
+                for i in range(len(dilation))
+            ]
+        )
+        self.convs2 = nn.ModuleList(
+            [
+                nn.Conv1d(
+                    channels,
+                    channels,
+                    kernel_size,
+                    stride=1,
+                    dilation=1,
+                    padding=self.get_padding(kernel_size, 1),
+                )
+                for _ in range(len(dilation))
+            ]
+        )
```
let's split this, 1 loop will be enough and add the conv to lists. Simpler and cleaner!

connor-henderson replied:
The whole HifiGan model was copied from `speecht5`, should I still make changes to both instances?

sanchit-gandhi replied:
Updating the SpeechT5 modelling code can go in a separate PR, then do a rebase on this PR and run `make fix-copies` to propagate the changes

IMO this refactor is not essential for the merging of this PR since it's actually SpeechT5 code that requires changing, but is certainly a nice bit of clean-up we can do. Feel free to open the PR if you have time for this @connor-henderson, otherwise let me know and I can open an issue so that we can keep track of this

connor-henderson replied:
@sanchit-gandhi I don't think I'll be able to get to it, opening an issue sounds good

ArthurZucker commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1759 @@
+    def get_padding(self, kernel_size, dilation=1):
+        return (kernel_size * dilation - dilation) // 2
+
+    def apply_weight_norm(self):
+        for layer in self.convs1:
+            nn.utils.weight_norm(layer)
+        for layer in self.convs2:
+            nn.utils.weight_norm(layer)
+
+    def remove_weight_norm(self):
+        for layer in self.convs1:
+            nn.utils.remove_weight_norm(layer)
+        for layer in self.convs2:
+            nn.utils.remove_weight_norm(layer)
```
Why do we need the `remove_weight_norm`. Also why not always apply the weight norm with: 
```python 
weight_norm = nn.utils.parametrizations.weight_norm(nn.Conv1d(...), name="weight", dim=2))
```
(split for clarity if needed 😉

connor-henderson replied:
Same as above, this module was identical between `FastSpeech2Conformer` and `SpeechT5` so I used `# Copied from ...`

sanchit-gandhi replied:
apply/remove weight norm is typically used for training/inference: https://github.com/huggingface/transformers/pull/24799#discussion_r1291505968

fine to keep it here imo

ArthurZucker reviewed (commented):
Wow that's a big model! Thanks for your hard work 💪🏻 I'll review once again later on, most of it is really clean. Modelling wise, there seems to be a lot of layer where we transpose on input and output which is a bit counter intuitive and adds overhead (if it can't be avoided then so be it!) 

The config is huge, with a loooot of arguments, good job on documenting all of this 🤗

connor-henderson commented on docs/source/fr/index.md:
```diff
@@ -108,6 +108,7 @@ La documentation est organisée en 5 parties:
 1. **[EncoderDecoder](model_doc/encoder-decoder)** (from Google Research) released with the paper [Leveraging Pre-trained Checkpoints for Sequence Generation Tasks](https://arxiv.org/abs/1907.12461) by Sascha Rothe, Shashi Narayan, Aliaksei Severyn.
 1. **[ERNIE](model_doc/ernie)** (from Baidu) released with the paper [ERNIE: Enhanced Representation through Knowledge Integration](https://arxiv.org/abs/1904.09223) by Yu Sun, Shuohuan Wang, Yukun Li, Shikun Feng, Xuyi Chen, Han Zhang, Xin Tian, Danxiang Zhu, Hao Tian, Hua Wu.
 1. **[ESM](model_doc/esm)** (from Meta AI) are transformer protein language models.  **ESM-1b** was released with the paper [Biological structure and function emerge from scaling unsupervised learning to 250 million protein sequences](https://www.pnas.org/content/118/15/e2016239118) by Alexander Rives, Joshua Meier, Tom Sercu, Siddharth Goyal, Zeming Lin, Jason Liu, Demi Guo, Myle Ott, C. Lawrenc ...
+1. **[FastSpeech2Conformer](model_doc/fastspeech2_conformer)** (from Microsoft Research and ESPnet) Microsoft Research released **FastSpeech2** with the paper [Fastspeech 2: Fast And High-quality End-to-End Text To Speech](https://arxiv.org/pdf/2006.04558.pdf) Yi Ren1, Chenxu Hu1, Xu Tan, Tao Qin, Sheng Zhao, Zhou Zhao, and Tie-Yan Liu. ESPnet built on this with a conformer version of FastSpeech2 ...
```
Maybe we should just mention the `espnet` paper then? Technically that's the only one we support, the msft one wasn't released

ArthurZucker replied:
Yes!

connor-henderson commented on src/transformers/modeling_outputs.py:
```diff
@@ -1412,6 +1412,51 @@ class Seq2SeqSpectrogramOutput(ModelOutput):
     """
     Base class for sequence-to-sequence spectrogram outputs.
 
+    Args:
```
Sounds good, how can I mark the breaking change?

ArthurZucker replied:
⚠️⚠️ something like this in the title of the PR 😉

ArthurZucker replied:
the best would be to not do this, but it's very limiting and a nice fix

connor-henderson replied:
Opted to remove the breaking change

connor-henderson commented on src/transformers/models/fastspeech2_conformer/convert_hifigan.py:
```diff
@@ -0,0 +1,136 @@
+logging.set_verbosity_info()
+logger = logging.get_logger("transformers.models.FastSpeech2Conformer")
+
+match_config = FastSpeech2ConformerHifiGan.from_pretrained("connor-henderson/fastspeech2_conformer_hifigan").config
```
Ah it's not, removed

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1759 @@
+        self.embed_dim = config.hidden_size
+        self.input_scale = math.sqrt(self.embed_dim)
+        self.dropout = nn.Dropout(p=module_config["positional_dropout_rate"])
+        self.pos_enc = None
```
Yes, ended up using the Llama implementation as per the below rec anyway

connor-henderson replied:
On second pass, using rotary embeddings instead actually causes a divergence so sticking with relative positional encoding. The reason the test was passing before was actually attributable to an env issue that caused old code to be used

connor-henderson commented on src/transformers/models/speecht5/modeling_speecht5.py:
```diff
@@ -3231,8 +3235,9 @@ def __init__(self, config: SpeechT5HifiGanConfig):
 
-        self.register_buffer("mean", torch.zeros(config.model_in_dim))
-        self.register_buffer("scale", torch.ones(config.model_in_dim))
+        if config.normalize_before:
+            self.register_buffer("mean", torch.zeros(config.model_in_dim))
+            self.register_buffer("scale", torch.ones(config.model_in_dim))
```
There's a couple other comments on the HifiGan below too - I didn't write it, just `# Copied from ...` given its the same HifiGan implementation as `SpeechT5`

connor-henderson commented on tests/models/fastspeech2_conformer/test_tokenization_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,166 @@
+        # fmt: on
+
+        self.assertListEqual(actual_encoding["input_ids"], expected_encoding["input_ids"])
+        self.assertListEqual(actual_encoding["attention_mask"], expected_encoding["attention_mask"])
```
We can't actually go from phoneme (id) to grapheme (decoded text) since its a one to many mapping

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1759 @@
+                Batch of input sequences.
+            padding_masks (`torch.ByteTensor` of shape `(batch_size, max_text_length)`, *optional*):
+                Batch of masks indicating padded part.
+            is_inference (`bool`, *optional*, defaults to `False`): Whether or not the model is running inference.
```
The only difference is that is_inference used whether labels are included or not to determine it's value, maybe I should change it to only self.training and raise an error if `self.training = True` but no labels are provided?

sanchit-gandhi replied:
To me, this later option is cleaner than introducing a new argument

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1759 @@
+        Args:
+            encoded_embeddings (`torch.Tensor` of shape `(batch_size, max_text_length, embedding_dim)`):
+                Batch of sequences of char or phoneme embeddings.
+            duration_labels (`torch.LongTensor` of shape `(batch_size, T)`):
+                Batch of durations of each frame.
```
Missed those, yes replaced

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1759 @@
+        )
+        self.dropout = nn.Dropout(module_config["dropout_rate"])
+
+    def forward(self, hidden_states):
```
removed the redundant transpose

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1759 @@
+        hidden_states = hidden_states.transpose(-1, 1)
+        hidden_states = self.conv1(hidden_states)
+        hidden_states = torch.relu(hidden_states)
+        hidden_states = hidden_states.transpose(-1, 1)
+        hidden_states = self.dropout(hidden_states)
+
+        hidden_states = hidden_states.transpose(-1, 1)
```
Per the below I don't think we have to actually, removed the intermediate redundant transposes

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1759 @@
+        # first compute matrix a and matrix c
+        # as described in https://arxiv.org/abs/1901.02860 Section 3.3
+        # (batch_size, head, time1, time2)
+        matrix_ac = torch.matmul(query_with_bias_u, key.transpose(-2, -1))
```
It was 4 dim where 1 and 2 swapped, then -1 and -2 so made it a one line permute instead

Commit:
address Arthur comments pt2

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1759 @@
+        return hidden_states
+
+
+class FastSpeech2ConformerAttention(nn.Module):
```
Updated

connor-henderson commented:
@ArthurZucker comments are addressed, left two follow up questions: first one on if changes requested on `# Copied from …` code in speecht5 hifigan should in fact be made, second one regarding self.training vs is_inference and checking for labels

Commit:
lint  changes

sanchit-gandhi commented:
Answered in-line! Thanks for iterating here @connor-henderson!

Commit:
Sanchit comment

connor-henderson commented:
Thanks for clarifying @sanchit-gandhi, addressed that. I believe the only lingering question is the one from this comment https://github.com/huggingface/transformers/pull/23439#discussion_r1302119798 around whether I should remove the tokenizer from the docstring examples to pass the doc tests since it requires g2p_en backend

ArthurZucker commented:
Hey! You can also add `g2p` to the list of package needed to run the PR documentation tests in the CI job! Do you need help with that?

sanchit-gandhi commented:
See @ydshieh's comment for reference: https://github.com/huggingface/transformers/pull/23439#discussion_r1333918558

Commit:
add g2p-en to doctest deps

Commit:
move up self.encoder

connor-henderson commented:
> Hey! You can also add `g2p` to the list of package needed to run the PR documentation tests in the CI job! Do you need help with that?

> See @ydshieh's comment for reference: https://github.com/huggingface/transformers/pull/23439#discussion_r1333918558

@sanchit-gandhi @ArthurZucker all good thanks was on vacation, just added

Commit:
onnx compatible tensor method

Commit:
fix is symbolic

sanchit-gandhi commented on src/transformers/models/fastspeech2_conformer/configuration_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,504 @@
+
+
+FASTSPEECH2_CONFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "connor-henderson/fastspeech2_conformer": "https://huggingface.co/connor-henderson/fastspeech2_conformer/raw/main/config.json",
```
Have requested access for you on the ESPNet organisation on the Hub so that you can move the checkpoints over and make edits to the model card - will keep you updated on its approval status!

sanchit-gandhi replied:
Alright we're both in! Feel free to move the weights at your convenience!

connor-henderson replied:
Great, I moved the weights over to ESPnet and have added model cards

sanchit-gandhi reviewed (approved):
Lovely - thanks again for iterating here @connor-henderson! Once @ArthurZucker gives the green light we can merge 👍

Commit:
fix paper url

connor-henderson commented on docs/source/en/model_doc/fastspeech2_conformer.md:
```diff
@@ -0,0 +1,58 @@
+
+## Overview
+
+The FastSpeech2Conformer model was proposed with the paper [Recent Developments On Espnet Toolkit Boosted By Conformer](https://arxiv.org/abs/2110.07205) by Pengcheng Guo, Florian Boyer, Xuankai Chang, Tomoki Hayashi, Yosuke Higuchi, Hirofumi Inaguma, Naoyuki Kamo, Chenda Li, Daniel Garcia-Romero, Jiatong Shi, Jing Shi, Shinji Watanabe, Kun Wei, Wangyou Zhang, and Yuekai Zhang.
```
Thanks just corrected

Commit:
move models to espnet org

Commit:
make style

Commit:
make fix-copies

Commit:
update docstring

ArthurZucker commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1696 @@
+        super().__init__(config)
+
+        self.model = FastSpeech2ConformerModel(config.model_config)
+        self.vocoder = FastSpeech2ConformerHifiGan(config.vocoder_config)
```
If base prefix is blocking you; I'll fix it

sanchit-gandhi replied:
Yep fine for me to use the AutoModel API like we do in MusicGen if it's not too much work for @connor-henderson to adapt the modelling code: https://github.com/huggingface/transformers/blob/e8fdd7875def7be59e2c9b823705fbf003163ea0/src/transformers/models/musicgen/modeling_musicgen.py#L1503

connor-henderson replied:
Base prefixing is blocking currently 
<img width="1303" alt="Screenshot 2023-10-10 at 1 43 28 PM" src="https://github.com/huggingface/transformers/assets/78612354/01389740-4ffd-4d24-85d4-2b92af0f06b0">

connor-henderson replied:
Maybe this could be in follow-on work to prevent it blocking this pr?

ArthurZucker replied:
Could you share the full traceback? 😉

ArthurZucker reviewed (approved):
🥳 Congrats, it's super clean. (apart from the module config! 👀 ), a great effort to match our API and add the training as well. 
One good addition would be to leverage the auto api for the vocoder. 
Otherwise LGTM

connor-henderson commented on src/transformers/models/auto/tokenization_auto.py:
```diff
@@ -25,7 +25,13 @@
 from ...tokenization_utils import PreTrainedTokenizer
 from ...tokenization_utils_base import TOKENIZER_CONFIG_FILE
-from ...utils import cached_file, extract_commit_hash, is_sentencepiece_available, is_tokenizers_available, logging
+from ...utils import (
```
haven't changed anything here just fyi, diff was from make style

connor-henderson replied:
Yes shouldn't have been on previous commit, but now it should be since `is_g2p_en_available` is now imported

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1696 @@
+    https://github.com/espnet/espnet/pull/2816. Paper: https://arxiv.org/abs/1901.02860.
+    """
+
+    def __init__(self, config: FastSpeech2ConformerConfig, module_config):
```
module_config comes from either the encoder or decoder config from `FastSpeech2ConformerConfig`, so we can't get to one config by overwriting

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1696 @@
+    """
+
+    def __init__(self):
+        super().__init__()
```
moved it to a fn

connor-henderson commented on src/transformers/models/fastspeech2_conformer/configuration_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,506 @@
+        super().__init__(**kwargs)
+
+    @classmethod
+    def from_sub_model_configs(
```
nope, removed

connor-henderson commented on docs/source/en/model_doc/fastspeech2_conformer.md:
```diff
@@ -0,0 +1,58 @@
+*Non-autoregressive text to speech (TTS) models such as FastSpeech (Ren et al., 2019) can synthesize speech significantly faster than previous autoregressive models with comparable quality. The training of FastSpeech model relies on an autoregressive teacher model for duration prediction (to provide more information as input) and knowledge distillation (to simplify the data distribution in output ...
+
+This model was contributed by [Connor Henderson](https://huggingface.co/connor-henderson). The original code can be found [here](https://github.com/espnet/espnet/blob/master/espnet2/tts/fastspeech2/fastspeech2.py).
+
```
Added

connor-henderson commented on src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py:
```diff
@@ -0,0 +1,1696 @@
+        dropout_rate=0.5,
+    ):
+        """
+        Initilize duration predictor module.
```
Just a variance predictor, edited

connor-henderson commented on docs/source/en/model_doc/fastspeech2_conformer.md:
```diff
@@ -0,0 +1,58 @@
+The abstract from the original FastSpeech2 paper is the following:
+
+*Non-autoregressive text to speech (TTS) models such as FastSpeech (Ren et al., 2019) can synthesize speech significantly faster than previous autoregressive models with comparable quality. The training of FastSpeech model relies on an autoregressive teacher model for duration prediction (to provide more information as input) and knowledge distillation (to simplify the data distribution in output ...
+
```
Added ex. snippet, by scheme do you mean screenshot of the architecture? haven't seen in the other model docs so not sure how to add, lmk what's wanted

ArthurZucker replied:
Yep a screen shot of the architecture. It's markdown so adding the image should work! It's not a must but the model is complicated so let's be helpful

connor-henderson replied:
Sounds good, added

Commit:
Arthur comments

Commit:
update docstring w/ new updates

Commit:
add model architecture images

Commit:
header size

Commit:
md wording update

Vaibhavs10 commented:
bump!

ArthurZucker commented:
I'll help to get this merge @connor-henderson as seen offline! I gotta finish catching up and should come back to this tomorrow ! Sorry again 🤗

ArthurZucker commented:
Here is [how I dealt with the `CLIPVisionModel` ](https://github.com/huggingface/transformers/pull/27662#discussion_r1415447950). In the PR I added the config and model to the config mappings and also the correct places to import, but not in the readmes or in the index file.

Commit:
make style

ylacombe commented:
It's been a long ride, but merging now!

Thanks again for the great work and your patience !
